docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,383 @@
+applies to mints and burns.
+
+Following this model, we never have to worry about compounding interest. We
+simply need to call the right program to perform instructions.
+
+## Required Changes
+
+A new interest-bearing token program is only one component of the overall solution,
+since i-tokens need to be easily integrated in on-chain programs, web apps,
+and wallets.
+
+### Token Program Conformance
+
+Since our solution entails the creation of a new token program, we're opening the
+floodgates for other app developers to do the same. To ensure safe and easy integration
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
+
+A new token program must include all of the same instructions as SPL token,
+but the "unchecked" variants should return errors. Concretely, this means the
+program must implement:
+
+* `InitializeMint`
+* `InitializeHolding` (previously `InitializeAccount`)
+* `Revoke`
+* `SetAuthority`
+* `CloseHolding` (previously `CloseAccount`)
+* `FreezeHolding` (previously `FreezeAccount`)
+* `ThawHolding` (previously `ThawAccount`)
+* `TransferChecked`
+* `ApproveChecked`
+* `MintToChecked`
+* `BurnChecked`
+* `InitializeHolding2` (previously `InitializeAccount2`)
+
+And the program should throw errors for:
+
+* `Transfer`
+* `Approve`
+* `MintTo`
+* `Burn`
+* `InitializeMultisig` (use a more general multisig)
+
+New instructions required:
+
+* `CreateHolding` (TODO name): only performs `Allocate` and `Assign` to self,
+useful for creating a holding when you don't know the program that you're
+interacting with. See the Associated Token Account section for how to use this.
+* `TransferAllChecked`: transfers everything from the source holding to the
+destination holding. Useful for closing a interest-bearing token holding since the
+balance is constantly changing.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `GetHoldingBalance`: balance of the holding
+* `GetHoldingOwner`: owner's public key
+* `GetMintSupply`: mint supply
+* `GetMintAuthority`: mint authority's public key
+* `GetMintDecimals`: mint decimals
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+The SPL token library provides wrappers compatible with any conforming program.
+For the base SPL token program, the wrapper simply deserializes the account data
+and returns the relevant data. For other programs, it calls the appropriate
+read-only instruction using an ephemeral account input.
+
+The structure for new holding and mint accounts must follow the layout of the
+existing SPL token accounts to ensure compatibility in RPC when fetching pre and
+post token balances.
+
+For a holding, the first 165 bytes must contain the same information
```

Comment:
Do we actually need to require that i-token holdings conform to any particular layout? As long as the program implements a `SerializeToSPL` method that is called anytime a caller wants to access the account as an SPL token, it seems like the underlying i-token holding could be whatever state is most efficient for the particular program.

For instance, in the specific case of interest-bearing tokens laid out below, there wouldn't be any need to store both share amount and a token amount (that may always be wrong). It could be instead:

```rust
pub struct Holding {
    // SPL Holding Fields
    pub mint: Pubkey,
    pub owner: Pubkey,
    pub share_amount: u64,
    pub delegate: COption<Pubkey>,
    pub state: AccountState,
    pub is_native: COption<u64>,
    pub delegated_amount: u64,
    pub close_authority: COption<Pubkey>,
}
```

where `SerializeToSPL` converts share_amount to current token_amount

Reply:
We'll need to expose SerializeToSPL over RPC as well for `spl-token` cli and other front-ends to use if we do this, not bad though.

I guess the counter is that we generally expect iTokens to fully implement SPL Token (and then add some), so why wouldn't they need all these fields.  Implementation wise I think we want to encourage a common SPL Token library crate that is built on rather than everybody forking from day 1, so we can easily publish updates for security fixes.  Not having a common `struct` to rally around makes integration of new SPL Token releases much harder for the downstream

Reply:
> We'll need to expose SerializeToSPL over RPC as well for `spl-token` cli and other front-ends to use if we do this, not bad though.

I think that we're going to need to do this regardless in order to return correct token amounts at any given time.

I see your point about new token programs being additive.
I suppose it is only 8 bytes. Just seems a waste, if amount is essentially going to become `unused`.

Reply:
I've updated the docs to enforce that new token programs conform their structs to look like SPL token, along with the important difference between UI amount and amount.